### ISSUE TITLE
empty training set: output/predict_misc/final_training_models.gff3

### ISSUE THREAD

**Issue by papelypluma** [nextgenusfs/funannotate#629]
Hi, I've encountered a similar issue referring to https://github.com/nextgenusfs/funannotate/issues/427 with funannotate predict. Quoting the following error below:

> [08/15/21 21:31:14]: Directory not copied. Error: [Errno 17] File exists: 'output/predict_misc/ab_initio_parameters/codingquarry'
> [08/15/21 21:31:14]: Snap training failed, empty training set: output/predict_misc/final_training_models.gff3
> [08/15/21 21:31:14]: snap failed removing from training parameters
> [08/15/21 21:31:14]: GlimmerHMM training failed, empty training set: output/predict_misc/final_training_models.gff3
> [08/15/21 21:31:14]: GlimmerHMM failed, removing from training parameters

This is a very similar error I got before re-running the prediction with:
> mkdir: cannot create directory ‘./ParameterFiles’: File exists

instead of "Directory not copied. ..."

Requesting for help to address this error. Thank you very much.

**Comment by papelypluma** [comment]
Hi @nextgenusfs. By any chance is there a quick fix or solution to address this issue? Thanks a lot!

**Comment by nextgenusfs** [comment]
I'm away from computer for the last week. The previous issue was fixed I thought. Is this only a problem on a rerun?

**Comment by nextgenusfs** [comment]
Is this still a problem? I need a reproducible test case or at least the commands you ran and in what sequence to try to fix.

**Comment by papelypluma** [comment]
Hi @nextgenusfs thanks for the help. I'm trying to re-run it again from the very beginning to check if the issue can be replicated. The train and predict steps were interrupted when I first ran it, but it seems that they can continue from where they left off. The issue first came out after codequarry step, so at that point I tried re-running predict again but still got the same issue. 

Now, I'm running predict to check  if this issue still persists.

**Comment by nextgenusfs** [comment]
@papelypluma thanks that will help.  Generally the pipeline can be re-run upon failure, but I could imagine a scenario where there are some intermediate files that get partial/incomplete written it isn't smart enough to not re-use them.

**Comment by papelypluma** [comment]
Hi @nextgenusfs. Thanks for the help. SNAP and Glimmerhmm have failed after codingquarry step. The pipeline is still running skipping these two from training parameters.

> [08/30/21 03:51:33]: Snap training failed, empty training set: output/predict_misc/final_training_models.gff3
> [08/30/21 03:51:33]: snap failed removing from training parameters
> [08/30/21 03:51:33]: GlimmerHMM training failed, empty training set: output/predict_misc/final_training_models.gff3
> [08/30/21 03:51:33]: GlimmerHMM failed, removing from training parameters

I've checked, and indeed the */final_training_models.gff3 is missing. 

I'm using funannotate v1.8.7 with the following 'predict' command:
> funannotate predict --input sample_cleaned_sorted_masked.fasta --out output --species "Genus species" --strain X --transcript_evidence alltx.okay.cds --protein_evidence hemiptera-16_orthodb.fasta $FUNANNOTATE_DB/uniprot_sprot.fasta --augustus_species rhodnius --cpus 16

**Comment by nextgenusfs** [comment]
Okay, so what is failing here is BUSCO then apparently.  What does the BUSCO log file say?

**Comment by papelypluma** [comment]
Doesn't seem to have any BUSCO log file, but just checked on augustus and it's v3.3.3 that came with the funannotate installation.

**Comment by nextgenusfs** [comment]
Can you attach the entire predict log so I can see what happened?

**Comment by papelypluma** [comment]
Here it is:

[funannotate-predict.log](https://github.com/nextgenusfs/funannotate/files/7073897/funannotate-predict.log)

**Comment by nextgenusfs** [comment]
Okay, since training data was available it tried to pull PASA GFF to use as training:
```
08/25/21 09:48:47]: /home/datboncan/miniconda3/envs/funannotate/bin/funannotate predict --input RpedSz_cleaned_sorted_masked.fasta --out output --species Riptortus pedestris --strain SZV --transcript_evidence alltx.okay.cds --protein_evidence hemiptera-16_orthodb.fasta /expt/datboncan/database/funannotate_db/uniprot_sprot.fasta --augustus_species rhodnius --cpus 16

[08/25/21 09:48:47]: OS: Ubuntu 18.04, 64 cores, ~ 528 GB RAM. Python: 3.7.8
[08/25/21 09:48:47]: Running funannotate v1.8.7
[08/25/21 09:48:47]: GeneMark path: /home/datboncan/tools/genemark/gm_et_linux_64
[08/25/21 09:48:48]: Full path to gmes_petap.pl: /home/datboncan/tools/genemark/gm_et_linux_64/gmes_petap.pl
[08/25/21 09:48:48]: GeneMark appears to be functional? True
[08/25/21 09:48:50]: Found training files, will re-use these files:
  --rna_bam output/training/funannotate_train.coordSorted.bam
  --pasa_gff output/training/funannotate_train.pasa.gff3
  --stringtie output/training/funannotate_train.stringtie.gtf
  --transcript_alignments output/training/funannotate_train.transcripts.gff3
[08/25/21 09:48:51]: {'augustus': 1, 'hiq': 2, 'genemark': 1, 'pasa': 6, 'codingquarry': 2, 'snap': 1, 'glimmerhmm': 1, 'proteins': 1, 'transcripts': 1}
[08/25/21 09:48:53]: {'augustus': 'pretrained', 'genemark': 'selftraining', 'snap': 'pasa', 'glimmerhmm': 'pasa', 'codingquarry': 'rna-bam'}
```
Is that file empty?
```
grep -c $'\tgene\t' output/training/funannotate_train.pasa.gff3
```

**Comment by nextgenusfs** [comment]
In other words, it should have just been copied over (I think):
`output/training/funannotate_train.pasa.gff3` should be the same as `output/predict_misc/final_training_models.gff3`.

**Comment by papelypluma** [comment]
It's not empty. Returned a value of 35048.

**Comment by nextgenusfs** [comment]
And that is the same as `output/predict_misc/final_training_models.gff3`?

**Comment by papelypluma** [comment]
That's missing in the output/predict_misc/ directory.

**Comment by nextgenusfs** [comment]
Does this file exist: `output/predict_misc/pasa_predictions.gff3`?

If you could just `ls -l the output/predict_misc` directory that works as well.

**Comment by papelypluma** [comment]
Yes, it exists.

> total 5.7G
> drwxrwxr-x 6 datboncan datboncan 4.0K Aug 30 03:51 ./
> drwxrwxr-x 6 datboncan datboncan 4.0K Aug 25 09:48 ../
> drwxrwxr-x 4 datboncan datboncan 4.0K Aug 30 03:51 ab_initio_parameters/
> -rw-rw-r-- 1 datboncan datboncan 5.8K Aug 25 09:55 assembly-gaps.bed
> -rw-rw-r-- 1 datboncan datboncan  46M Aug 27 05:12 augustus.evm.gff3
> -rw-rw-r-- 1 datboncan datboncan  46M Aug 27 05:12 augustus.evm.gff3.bak
> -rw-rw-r-- 1 datboncan datboncan  86M Aug 27 05:11 augustus.gff3
> -rw-rw-r-- 1 datboncan datboncan 7.7M Aug 25 10:09 bam_hints.joined.tmp
> -rw-rw-r-- 1 datboncan datboncan 7.7M Aug 25 10:09 bam_hints.sorted.tmp
> -rw-rw-r-- 1 datboncan datboncan 7.7M Aug 25 10:09 bam_hints.tmp
> drwxrwxr-x 5 datboncan datboncan 4.0K Aug 30 03:50 CodingQuarry/
> -rw-rw-r-- 1 datboncan datboncan 157M Aug 30 03:51 coding_quarry.gff3
> drwxrwxr-x 4 datboncan datboncan 4.0K Aug 30 09:39 EVM/
> drwxrwxr-x 6 datboncan datboncan 4.0K Aug 26 23:16 genemark/
> -rw-rw-r-- 1 datboncan datboncan 266M Aug 26 23:18 genemark.evm.gff3
> -rw-rw-r-- 1 datboncan datboncan 266M Aug 26 23:18 genemark.evm.gff3.bak
> -rw-rw-r-- 1 datboncan datboncan 292M Aug 26 23:17 genemark.gff
> -rw-rw-r-- 1 datboncan datboncan 266M Aug 26 23:18 genemark.temp.gff
> -rw-rw-r-- 1 datboncan datboncan 499M Aug 30 03:51 gene_predictions.gff3
> -rw-rw-r-- 1 datboncan datboncan 1.1G Aug 25 09:55 genome.softmasked.fa
> -rw-rw-r-- 1 datboncan datboncan 957K Aug 26 22:36 gmhmm.mod
> -rw-rw-r-- 1 datboncan datboncan 123M Aug 26 18:38 hints.ALL.gff
> -rw-rw-r-- 1 datboncan datboncan 279M Aug 26 18:37 hints.all.sort.tmp
> -rw-rw-r-- 1 datboncan datboncan 279M Aug 26 18:35 hints.all.tmp
> -rw-rw-r-- 1 datboncan datboncan 6.2M Aug 25 10:09 hints.BAM.gff
> -rw-rw-r-- 1 datboncan datboncan 271M Aug 25 10:08 hints.M.gff
> -rw-rw-r-- 1 datboncan datboncan 2.0M Aug 26 18:35 hints.P.gff
> -rw-rw-r-- 1 datboncan datboncan  52M Aug 26 18:35 p2g.diamond.out
> -rw-rw-r-- 1 datboncan datboncan  32M Aug 25 09:48 pasa_predictions.gff3
> -rw-rw-r-- 1 datboncan datboncan 2.0M Aug 26 18:35 protein_alignments.gff3
> -rw-rw-r-- 1 datboncan datboncan 323M Aug 25 10:09 proteins.combined.fa
> -rw-rw-r-- 1 datboncan datboncan 170M Aug 25 09:55 repeatmasker.bed
> -rw-rw-r-- 1 datboncan datboncan  64M Aug 27 05:12 stringtie.gff3
> -rw-rw-r-- 1 datboncan datboncan 244M Aug 25 10:08 transcript_alignments.gff3
> -rw-rw-r-- 1 datboncan datboncan 129M Aug 25 10:05 transcript_evidence_unique.bam
> -rw-rw-r-- 1 datboncan datboncan 622M Aug 25 09:56 transcript_evidence_unique.fasta
> -rw-rw-r-- 1 datboncan datboncan 165M Aug 25 10:07 transcript_evidence_unique.gff3
> -rw-rw-r-- 1 datboncan datboncan  181 Aug 30 03:51 weights.evm.txt

**Comment by nextgenusfs** [comment]
Oh okay, this is same as https://github.com/nextgenusfs/funannotate/issues/591. Fixed in v1.8.9.
